llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp

```diff
@@ -1390,14 +1390,34 @@ InstructionCost RISCVTTIImpl::getMemoryOpCost(unsigned Opcode, Type *Src,
   InstructionCost Cost = 0;
   if (Opcode == Instruction::Store && OpInfo.isConstant())
     Cost += getStoreImmCost(Src, OpInfo, CostKind);
-  InstructionCost BaseCost =
-    BaseT::getMemoryOpCost(Opcode, Src, Alignment, AddressSpace,
-                           CostKind, OpInfo, I);
+
+  std::pair<InstructionCost, MVT> LT = getTypeLegalizationCost(Src);
+
+  InstructionCost BaseCost = [&]() {
+    InstructionCost Cost = LT.first;
+    if (CostKind != TTI::TCK_RecipThroughput)
+      return Cost;
+
+    // Our actual lowering uses a VL predicated load of the next legal type,
+    // or splitting if there is no wider legal type.  This is reflected in
+    // the result of getTypeLegalizationCost, but BasicTTI assumes the
+    // widened cases are scalarized.
+    const DataLayout &DL = this->getDataLayout();
+    if (Src->isVectorTy() && LT.second.isVector()) {
+      auto *SrcVT = cast<VectorType>(Src);
+      TypeSize SrcEltSize = DL.getTypeStoreSizeInBits(SrcVT->getElementType());
+      TypeSize LegalEltSize = LT.second.getVectorElementType().getSizeInBits();
+      if (SrcEltSize == LegalEltSize)
```

Comment:
I presume the case this is trying to catch is <n x i64> on zve32x, but wouldn't that be expanded not promoted?

Reply:
The primary motivation is <3 x i32> on V.  The scalable vector thing on zve32x was merely a side finding.  And no, that would be a single vle on the wider legal type.  Here's a standalone test which shows that:

```
; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
; RUN: llc -mtriple=riscv64 -mattr=+zve32x < %s | FileCheck %s

define void @test(ptr %a, ptr %b) nounwind vscale_range(4,1024) {
; CHECK-LABEL: test:
; CHECK:       # %bb.0:
; CHECK-NEXT:    csrr a2, vlenb
; CHECK-NEXT:    srli a2, a2, 3
; CHECK-NEXT:    vsetvli zero, a2, e32, m1, ta, ma
; CHECK-NEXT:    vle32.v v8, (a0)
; CHECK-NEXT:    vse32.v v8, (a1)
; CHECK-NEXT:    ret
  %LD = call <vscale x 1 x i32> @llvm.masked.load.nxv1i32.p0(ptr %a, i32 4, <vscale x 1 x i1> splat(i1 true), <vscale x 1 x i32> poison)
  store <vscale x 1 x i32> %LD, ptr %b
  ret void
}

```

Reply:
Handling the case where <3 x i32> is widening makes sense to me, but I think I'm missing something with the `SrcEltSize == LegalEltSize` check. When will a vector's element type get legalized to different size? Won't the two always be the same?

Reply:
When writing an updated description for this review which was inclusive of the splitting behavior in the cost model, I sat down to write a couple of lowering tests to be sure the costing I was describing was both correct and matched actual lowering.

When doing so, I found a test case which crashes during lowering (see below).  Given that, I added a restriction to this change to specifically only change the cost when promoting.  This still covers the cases I'm interesting in, and avoids the possibility of perturbing the costs in an area where we appear to have a lowering problem.

```
; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
; RUN: llc -mtriple=riscv64 -mattr=+zve32x < %s | FileCheck %s

; Note that vscale x 32 x i32 appears to work just fine, and we split to two LMUL 8
; types as expected.  It's apparently splitting combined with the need to lower the
; second half with VL predication which is causing problems.
define void @test31(ptr %a, ptr %b) nounwind vscale_range(4,1024) {
  %LD = call <vscale x 31 x i32> @llvm.masked.load.nxv31i32.p0(ptr %a, i32 4, <vscale x 31 x i1> splat(i1 true), <vscale x 31 x i32> poison)
  store <vscale x 31 x i32> %LD, ptr %b
  ret void
}




```

Reply:
> Handling the case where <3 x i32> is widened makes sense to me, but I think I'm missing something with the `SrcEltSize == LegalEltSize` check. When will a vector's element type get legalized to different size? Won't the two always be the same?

I'd originally asserted this locally, and the assert failed fairly widely.  I didn't dig into it in any detail, but I think it involved a vector of i1.

Reply:
I checked out this branch locally and added in the assert, it didn't seem to trigger. Maybe something changed in the meantime? We have legal i1 vector types so I presume if it was to get legalized the element type would still be i1.

Reply:
For the record I'm happy to leave this in defensively. Everything else LGTM

Reply:
Thanks for pushing on this.  I went back and ran the local tests with the assert, and found out that my assert/check was actually entirely spurious.  

The case that I found locally (without the isKnownLE check) is:
```
<2 x i1> 
v2i1 
```

After staring at that for a moment, and going "huh?", I realized that the check was using the Store size of the element from DL, and the non-store width of the element from the VT.  Fixing that mismatch causes the assert to never fail.  

Given that, I'm just going to remove the check entirely.